body.tex

```diff
@@ -1,166 +1,466 @@
+
+The performance penalty of the behaviour is debilitating. Bellow is an abreviated query plan for the simplest case of \lstinline[basicstyle=\ttfamily]{select * from visit_detector_patch_join limit 4;} that suffers from the described issue.
+
+\begin{lstlisting}[style=sqlprompt]
+Limit .... (actual time=12732.766..12732.774 rows=4 loops=1)
+->Unique .... (actual time=12732.765..12732.770 rows=4 loops=1)
+  ->Sort .... (actual time=12732.764..12732.767 rows=6 loops=1)
+    Sort Key: ....
+    Sort Method: external merge  Disk: 209008kB
+    ->Hash Join .... (actual time=763.846..1524.504 rows=4642107 loops=1)
+      Hash Cond: ....
+      ->Seq Scan on .... (actual time=0.008..16.289 rows=206960 loops=1)
+      ->Hash .... (actual time=763.723..763.724 rows=3259107 loops=1)
+        Buckets: 65536  Batches: 64  Memory Usage: 3635kB
+        ->  Seq Scan on patch_skypix_join .... (actual time=0.006..240.128 rows=3259107 loops=1)
+Planning time: 0.323 ms
+Execution time: 12759.391 ms
+\end{lstlisting}
+
+Note that the total materialized size of the view on disk is 200MB and that it took 12 seconds to retrieve only 4 results! There is no clear solution to this problem from within the Middleware codebase. 
+
+The sollution is however possible by manually redefining the views as materialized views instead, adding triggers that recreate the views on any insert statement to the underlying tables that make the view, and then adding indexing onto the materialized views. This reduces the time it takes to create a QuantumGraph to the point where it's comparable to that reported by the DM team. The query plan for the same SQL select statement as described above now looks like:
+
+\begin{lstlisting}[style=sqlprompt]
+Limit .... (actual time=0.009..0.010 rows=4 loops=1)
+->Seq Scan on visit_detector_patch_join .... (actual time=0.008..0.008 rows=4 loops=1)
+Planning time: 0.053 ms
+Execution time: 0.021 ms
+\end{lstlisting}
+
+Recently \jira{DM-17023} reworked and completely re-implemented the SQL schema and data model for the registries. We have not yet tested the performance of the new registry schema in PostgreSQL. The issues described here well optimize the set of specific queries, at the cost of a performance hit for all inserts into visits, patches..., but an important take away lesson here should be that because all of the SQL is generated completely dynamically guaranteeing execution plan stability and performance is dificult. Not all DBMSs optimize SQL statements equally well. Tying the CI and testing to only one database doesn't suficiently test the code performance and because the SQL is generated completely dynamically to fix potential poor performance requires schema modifications to be made external to the Middleware code. Because some of the performance loss can be debilitating, it is possible that certain DBMSs are, in this implementation, tied into implementing unoptimal solutions just to be able to run DRP.
```

Comment:
The dynamic SQL is always of a particular form.  I don't think we have shown that there are _data-dependent_ performance problems.  RDBMS optimization difficulties and differences between RDBMSes apply to static SQL as well as dynamic SQL.  I would rewrite this paragraph.

Reply:
Not sure the comment relates, or maybe I don't understand it. The claim is that because the SQL statements are created dynamically, changes in the code-base affects the statements. A lot of work and effort (multiple weeks) was exerted in optimizing the queries in w38 build. That work is inconsequential for w39 build. Granted in w39 a lot changed, but that doesn't need to be the case. If someone changed a single line of code, that changes the order of select statements or order of conditions in it, just that could invalidate the work that went into getting it to work. These changes could come even outside from the middleware - by lets say SQLAlchemy changing the internals of how they produce the SQL statements. 

Dynamically created statements are perpetually the same, I agree. I don't know how data changes could affect the registry drastically (since most of the data in it is a simple int or string anyhow). The performance is very dependent on how that code produces the statements though and it takes a lot of effort to optimize it. It could also sometimes be impossible in itself, or at least will bring penalties on some other front (f.e. materialized views have triggers on visit inserts).

Reply:
I rewrote this paragraph based on what I understand from here.